src/kw/kw.cpp

```diff
@@ -21,123 +25,183 @@
-    if(curl) {
+    const std::string body = "{\n\"username\": \"" + username +
+        "\", \n\"password\": \"" + password + "\"\n}";
 
-        // TODO: use json library
-        string payload = "{\n\"username\": \"" + username + "\", \n\"password\": \"" + password + "\"\n}";
+    std::shared_ptr<JSON_Value> res;
 
-        // Apply API expected headers
-        chunk = curl_slist_append(chunk, "Accept: application/json");
-        chunk = curl_slist_append(chunk, "Content-Type: application/json");
-        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, chunk);
+    try {
+        res = this->http_client->POST(
+            "https://worldapi-dev.kano.me/accounts/auth",
+            body);
+    } catch (const HTTPRequestFailedError& err) {
+        if (verbose) {
+            std::cout << err.what() << std::endl;
+        }
+
+        // TODO: Do we actually want to log out in this scenario?
+        this->logout();
+        return false;
+    } catch (const SessionInitError& err) {
+        if (verbose) {
+            std::cout << err.what() << std::endl;
+        }
+
+        // TODO: Do we actually want to log out in this scenario?
+        this->logout();
+        return false;
+    } catch (const std::exception& err) {
+        if (verbose) {
+            std::cout << "Unkown error: " << err.what() << std::endl;
+        }
 
-        curl_easy_setopt(curl, CURLOPT_URL, "https://worldapi-dev.kano.me/accounts/auth");
-        curl_easy_setopt(curl, CURLOPT_POST, 1L);
+        // TODO: Do we actually want to log out in this scenario?
+        this->logout();
+        return false;
+    }
 
-        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYPEER, 0L);
-        curl_easy_setopt(curl, CURLOPT_SSL_VERIFYHOST, 0L);
-        curl_easy_setopt(curl, CURLOPT_VERBOSE, verbose ? 1L : 0L);
+    this->token = this->parse_token(res);
+    this->expiration_date = this->parse_expiration_date(res);
+    this->save_data();
+
+    if (verbose) {
+        std::shared_ptr<char> resp_str(
+            json_serialize_to_string(res.get()),
+            json_free_serialized_string);
+        std::cout << ">>> login SERVER RESPONSE: " << resp_str << std::endl;
+        std::cout << ">>> Token: " << get_token() << std::endl;
+        std::cout << ">>> Expiration date: " << get_expiration_date()
+                  << std::endl;
+    }
 
-        curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, payload.length());
-        curl_easy_setopt(curl, CURLOPT_COPYPOSTFIELDS, payload.c_str());
+    return true;
+}
 
-        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_function);
 
-        /* Perform the request, res will get the return code */
-        res = curl_easy_perform(curl);
+/**
+ * \brief Logs out the current user
+ *
+ * The cache file is removed so that the login details are discarded
+ *
+ * \param verbose     Be more descriptive on the transaction details
+ *
+ * \returns True if logout was successful
+ */
+bool KanoWorld::logout(const bool verbose)
```

Comment:
Candidate for:
```suggestion
bool KanoWorld::logout(const bool verbose) const
```

Reply:
Actually, realised that we need to clear the class variables for this guy so not gonna be `const`.